Add delete endpoint for webhooks, optionally deleting PipelineRuns

a-roberts opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Adds the ability to delete webhooks by name and if &deletepipelineruns=true is set, delete the PipelineRuns that are for the same Git repository URL. This is only going to work for PipelineRuns we create through the extension as we label them. These labels are what I use to do the comparison with.

I've also made the images Alpine to reduce their size (for our tests Dockerfiles), added tests and API docs.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

a-roberts commented:
@ncskier @dibbles can you please take a look?

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+		if err != nil {
+			theError := errors.New("bad request information provided, cannot handle deletepipelineruns query (should be set to true or not provided)")
+			logging.Log.Error(theError)
+			RespondError(response, err, http.StatusInternalServerError)
```
I think this should `return` after responding with an error message

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+	if namespace == "" {
+		theError := errors.New("no namespace was provided")
+		logging.Log.Error(theError)
+		RespondError(response, theError, http.StatusBadRequest)
```
I think this should `return` after responding with an error message

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+		if err != nil {
+			if strings.Contains(err.Error(), "could not find webhook with name") {
+				logging.Log.Errorf("webhook (name %s) not found in namespace %s", name, namespace)
+				RespondError(response, err, http.StatusNotFound)
```
I think this should `return` after responding with an error message

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+				RespondError(response, err, http.StatusNotFound)
+			} else {
+				logging.Log.Errorf("error deleting the webhook (name %s), error: %s", name, err)
+				RespondError(response, err, http.StatusInternalServerError)
```
I think this should `return` after responding with an error message

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+		err = r.deleteWebhookFromConfigMapByName(name, namespace)
+		if err != nil {
+			logging.Log.Errorf("error deleting the webhook information (name %s) from the ConfigMap, error: %s", name, err)
+			RespondError(response, err, http.StatusInternalServerError)
```
I think this should `return` after responding with an error message

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+		}
+	} else {
+		logging.Log.Error("no name was provided")
+		RespondError(response, err, http.StatusBadRequest)
```
I think this should `return` after responding with an error message

ncskier commented on webhooks-extension/endpoints/webhook_test.go:
```diff
@@ -194,13 +209,273 @@ func testGetAllWebhooks(expectedWebhooks []webhook, r *Resource, t *testing.T) {
+
+func TestDeleteByNameNoNamespaceBadRequest(t *testing.T) {
+	setUpServer()
+	httpReq, _ := http.NewRequest(http.MethodDelete, server.URL+"/webhooks/name=foo", nil)
```
I think `webhooks/name=foo` should be `webhooks/foo` because the name is a path parameter.

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+	if name != "" {
+		foundRepoURL := r.findRepoURLFromConfigMap(name, namespace)
+		// This will remove any PipelineRuns too if specified to, hence the need for the repository URL
+		err = r.deleteGitHubWebhookByName(name, namespace, foundRepoURL, toDeletePipelineRuns)
```
If I'm not mistaken, all of the Github webhooks are created in the install namespace (this is because they need to be in the same namespace as their sink). So, I think this `deleteGitHubWebhookByName` should exclusively delete from the install namespace.

dibbles replied:
I would partly argue for leaving in the namespace arg as giving more future potential ..... it also allows for deleting when a sink is in a different namespace to the install namespace??  Depends on what install namespace is defined as .... presumably the install namespace of the dashboard

a-roberts replied:
I'll leave in for future potential as we want to work with various namespaces eventually

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+			"namespace":"default",
+			"gitrepositoryurl":"https://myrepourl",
+		  "accesstoken":"my-secret-for-hooks",
+		  "pipeline":"simple-helm-pipeline-insecure",
```
nit: please indent these two lines to match the others.

dibbles commented on webhooks-extension/cmd/extension/README.md:
````diff
@@ -55,6 +55,24 @@ Example POST
+### DELETE endpoint
+
+```
+DELETE /webhooks/<webhookid>?namespace=<my namespace>
````
interesting design question here - do we want namespace as a query param or as a path param ..... namespace/{namespace}/webhooks/{webhookid} ...... although we quite like having the /webhooks start ... so maybe /webhooks/namespace/{namespace}/webhooks/{webhook} although that seems long winded ... maybe query param is the nicest.

we get into quite a deep conversation about differences between the dashboard and the extension ... the dashboard being designed so that you choose a namespace and then all controls are on that namespace .... vs the extension taking namespace as selection in many places.  Possibly we should discuss as a wider group

a-roberts replied:
I'll leave in the query namespace for now for future proofing, we can behave a little differently from the main dashboard with our own API definitions and, like you, I'm quite a fan of the namespace being a query parameter, so I'm going to leave it as is. I don't really want to change everything to be, for example, /v1/namespaced, at least not yet, until we have a proper namespace and consistency story going forward

dibbles commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+				RespondError(response, err, http.StatusInternalServerError)
+			}
+		} else {
+			logging.Log.Info("Deleted the webhook OK, deleting from ConfigMap next")
```
%s webhook name to identify which one was deleted ..... might also want to add similar to method entry after reading values.

dibbles commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+
+	itsANestedMapAsBytes, err := json.Marshal(itsANestedMap)
+	if err != nil {
+		logging.Log.Errorf("Couldn't go back to bytes from the nested map, error is: %s", err.Error())
```
nit: lower case error message

dibbles commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+
+	configMap.BinaryData["GitHubSource"] = itsANestedMapAsBytes
+
+	_, err = configMapClient.Update(configMap)
```
discussion: does there need to be some locking around updates to the config map?

a-roberts replied:
I agree, added a test case, need to fix implementation next.

If two users are attempting to update the configmap at the same time, we may end up with the wrong data (let's say Adam does a delete to remove his webhook as Duane does the same thing to remove theirs, and we're behaving particularly slow). Duane's update is going to work on the old ConfigMap's data.

https://golang.org/src/net/http/client.go the http client itself that we use mentions being thread safe...

I've added a simple test case using a WaitGroup with two goroutines, both doing requests to delete different webhooks, and then we check they were really both deleted OK (that is, one didn't act upon some invalid data). This should simulate two HTTP requests interacting with the same resource. I thought all was well but on wrapping said test in a for loop and doing it 100 times, there's a 1/10 chance (observed so far!) that sometimes not all of the data is removed successfully. So I'll add a lock.

a-roberts commented:
@dibbles @ncskier please take another look, approve and lgtm if you're happy, I've addressed the feedback - keeping in the existing namespace behaviour for future proofing, adding a test to ensure it's thread safe (it wasn't, added a mutex)

ncskier commented:
Good catch with the lock @dibbles. @a-roberts I think we'll also need to lock here: https://github.com/tektoncd/experimental/blob/master/webhooks-extension/endpoints/webhook.go#L122-L129 to protect the read, modify, write code.

a-roberts commented:
Added, and I've also made the response be http 200 instead of 201 created

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -19,17 +19,43 @@ import (
 )
 
+var statusLock sync.Mutex
+var createLock sync.Mutex
```
Thanks for adding the extra lock, but I think there has to be one lock for both of them instead of two locks 😄 Otherwise, we still have the same problem if a CREATE and DELETE operation happen at the same time.

a-roberts replied:
I agree, so I've renamed it to modifyingConfigMapLock and I reference the same lock in both the delete and create functions now (I'm happy `defer` is a thing and hoping I needn't manually manage the lock with this approach).

I added another two test cases.

One for doing lots of creations at once with multiple routines to check the ConfigMap integrity.

Another is for create and delete.

@ncskier please take another look.

ncskier replied:
Oh, wow, that is a lot of logs 😅 I'm not sure how to limit the logs either. If it's not worth removing that `Debugf` call, then I think it's alright. I like that the tests are covering all of these sync issues 👍

a-roberts replied:
IKR, it's good to know what the webhooks are at the exact point of doing this call though.

I'd really prefer it if we could just turn off the logger and I don't want less iterations for obvious reasons (sometimes the tests will pass by chance).

Alternatively, sending stdout to /dev/null for these go tests as I've tried here. We test with `go test -v` and if I remove that verbose flag it's like you never ran the tests.

ncskier commented on webhooks-extension/endpoints/webhook.go:
```diff
@@ -126,12 +154,153 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+// Delete the webhook information from our ConfigMap
+func (r Resource) deleteWebhookFromConfigMapByName(webhookName, namespace string) error {
+	// We don't want any other requests going through here so use a mutex
+	// otherwise we'll end up with operations performed on old configmaps so we'll have invalid data
```
I think these comments are leftover from before you moved the lock code. They can probably be moved/removed now?

a-roberts replied:
Yep can go, done

ncskier commented:
@a-roberts I just re-tested your code, and it worked well 🙂Just have that last request about the stale comments, and should be good to go!

ncskier commented:
Thanks Adam!
/lgtm
